File: core/src/main/kotlin/me/vihara/atemwrapper/core/EventManager.java
package me.vihara.atemwrapper.core;

import me.vihara.atemwrapper.api.device.event.AtemEvent;
import me.vihara.atemwrapper.api.device.event.AtemEventListener;
import me.vihara.atemwrapper.api.device.event.AtemEventManager;
import me.vihara.atemwrapper.api.device.event.AtemEventHandler;

import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicInteger;

public final class EventManager implements AtemEventManager {
    public static final EventManager INSTANCE = new EventManager();

    private final Map<Integer, List<ListenerWrapper>> priorityGroups = new ConcurrentHashMap<>();
    private final Map<AtemEventListener, ListenerWrapper> listenerMap = new ConcurrentHashMap<>();
    private static final Map<Class<?>, List<Method>> METHOD_CACHE = new ConcurrentHashMap<>();
    private final AtomicInteger maxPriority = new AtomicInteger(Integer.MIN_VALUE);
    private final AtomicInteger minPriority = new AtomicInteger(Integer.MAX_VALUE);

    @Override
    public void addListener(final AtemEventListener listener) {
        final int priority = getListenerPriority(listener);
        final ListenerWrapper wrapper = new ListenerWrapper(listener, priority);

        listenerMap.put(listener, wrapper);
        priorityGroups.computeIfAbsent(priority, k -> new ArrayList<>()).add(wrapper);

        maxPriority.updateAndGet(current -> Math.max(current, priority));
        minPriority.updateAndGet(current -> Math.min(current, priority));
    }

    @Override
    public void removeListener(final AtemEventListener listener) {
        final ListenerWrapper wrapper = listenerMap.remove(listener);
        if (wrapper != null) {
            priorityGroups.computeIfPresent(wrapper.priority, (k, list) -> {
                list.remove(wrapper);
                return list.isEmpty() ? null : list;
            });
        }
    }

    @Override
    public void fireEvent(final AtemEvent event) {
        for (int priority = maxPriority.get(); priority >= minPriority.get(); priority--) {
            final List<ListenerWrapper> listeners = priorityGroups.get(priority);
            if (listeners != null) {
                for (int i = 0; i < listeners.size(); i++) {
                    listeners.get(i).invoke(event);
                }
            }
        }
    }

    private static int getListenerPriority(final AtemEventListener listener) {
        return METHOD_CACHE.computeIfAbsent(listener.getClass(), EventManager::findMethods)
                .stream()
                .findFirst()
                .map(method -> method.getAnnotation(AtemEventHandler.class).priority())
                .orElse(0);
    }

    private static List<Method> findMethods(Class<?> clazz) {
        final List<Method> methods = new ArrayList<>();
        for (final Method method : clazz.getMethods()) {
            if (method.isAnnotationPresent(AtemEventHandler.class)) {
                method.setAccessible(true);
                methods.add(method);
            }
        }
        return methods;
    }

    private static final class ListenerWrapper {
        private final AtemEventListener listener;
        private final int priority;
        private final List<Method> cachedMethods;

        ListenerWrapper(final AtemEventListener listener, final int priority) {
            this.listener = listener;
            this.priority = priority;
            this.cachedMethods = METHOD_CACHE.computeIfAbsent(listener.getClass(), EventManager::findMethods);
        }

        void invoke(final AtemEvent event) {
            for (int i = 0; i < cachedMethods.size(); i++) {
                final Method method = cachedMethods.get(i);
                try {
                    if (method.getParameterCount() == 1 &&
                            method.getParameterTypes()[0].isAssignableFrom(event.getClass())) {
                        method.invoke(listener, event);
                    }
                } catch (ReflectiveOperationException e) {
                    throw new RuntimeException("Failed to invoke event handler", e);
                }
            }
        }
    }
}

File: core/src/main/kotlin/me/vihara/atemwrapper/core/EventManager.kt
package me.vihara.atemwrapper.core

import me.vihara.atemwrapper.api.event.AtemEvent
import me.vihara.atemwrapper.api.event.AtemEventListener
import me.vihara.atemwrapper.api.event.manager.AtemEventManager
import me.vihara.atemwrapper.api.event.AtemEventHandler
import java.lang.reflect.Method
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicInteger

object EventManager : AtemEventManager {
    private val priorityGroups: MutableMap<Int, MutableList<ListenerWrapper>> = ConcurrentHashMap()
    private val listenerMap: MutableMap<AtemEventListener, ListenerWrapper> = ConcurrentHashMap()
    private val METHOD_CACHE: MutableMap<Class<*>, List<Method>> = ConcurrentHashMap()
    private val maxPriority = AtomicInteger(Int.MIN_VALUE)
    private val minPriority = AtomicInteger(Int.MAX_VALUE)

    override fun addListener(listener: AtemEventListener) {
        val priority = getListenerPriority(listener)
        val wrapper = ListenerWrapper(listener, priority)

        listenerMap[listener] = wrapper
        priorityGroups.computeIfAbsent(priority) { mutableListOf() }.add(wrapper)

        maxPriority.updateAndGet { current -> maxOf(current, priority) }
        minPriority.updateAndGet { current -> minOf(current, priority) }
    }

    override fun removeListener(listener: AtemEventListener) {
        listenerMap.remove(listener)?.let { wrapper ->
            priorityGroups.computeIfPresent(wrapper.priority) { _, list ->
                list.remove(wrapper)
                if (list.isEmpty()) null else list
            }
        }
    }

    override fun fireEvent(event: AtemEvent) {
        for (priority in maxPriority.get() downTo minPriority.get()) {
            val listeners = priorityGroups[priority]
            listeners?.forEach { it.invoke(event) }
        }
    }

    private fun getListenerPriority(listener: AtemEventListener): Int {
        return METHOD_CACHE.computeIfAbsent(listener::class.java, ::findMethods)
            .firstOrNull()
            ?.getAnnotation(AtemEventHandler::class.java)
            ?.priority
            ?: 0
    }

    private fun findMethods(clazz: Class<*>): List<Method> {
        return clazz.methods.filter { it.isAnnotationPresent(AtemEventHandler::class.java) }.onEach { it.isAccessible = true }
    }

    private class ListenerWrapper(
        private val listener: AtemEventListener,
        val priority: Int
    ) {
        private val cachedMethods: List<Method> = METHOD_CACHE.computeIfAbsent(listener::class.java, ::findMethods)

        fun invoke(event: AtemEvent) {
            for (method in cachedMethods) {
                try {
                    if (method.parameterCount == 1 && method.parameterTypes[0].isAssignableFrom(event::class.java)) {
                        method.invoke(listener, event)
                    }
                } catch (e: ReflectiveOperationException) {
                    throw RuntimeException("Failed to invoke event handler", e)
                }
            }
        }
    }
}
